third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -867,6 +817,134 @@ class BlockedToWMMA : public OpRewritePattern<tt::DotOp> {
+        return false;
+    }
+    return true;
+  }
+
+  LogicalResult matchAndRewrite(DotOp dotOp,
+                                PatternRewriter &rewriter) const override {
+    auto a = dotOp.getA();
+    auto b = dotOp.getB();
+    auto c = dotOp.getC();
+    auto d = dotOp.getD();
+
+    if (!isa<BlockedEncodingAttr>(d.getType().getEncoding()))
+      return failure();
+
+    Type aElTy = a.getType().getElementType();
+    Type bElTy = b.getType().getElementType();
+    Type cElTy = c.getType().getElementType();
+    Type dElTy = d.getType().getElementType();
+
+    int rank = a.getType().getShape().size();
+    int k = a.getType().getShape()[rank - 1];
+
+    bool dotAvailable = arch == "gfx908" || arch == "gfx90a" ||
+                        arch.starts_with("gfx94") ||
+                        arch.starts_with("gfx11") || arch.starts_with("gfx103");
+
+    // Try Fp16 x Fp16 -> Fp32 dot
+    if (dotAvailable && aElTy.isF16() && bElTy.isF16() && cElTy.isF32() &&
+        dElTy.isF32()) {
+      if (k % 2 == 0) {
+        // nothing to do for this dot
+        return failure();
+      }
+      // if k % 2 != 0: can not use DOT instruction, continue with FMA
+    }
+    // Try I8 x I8 -> I32 dot
+    if (dotAvailable && aElTy.isInteger(8) && bElTy.isInteger(8) &&
+        cElTy.isInteger(32) && dElTy.isInteger(32)) {
+      if (k % 4 == 0) {
+        // nothing to do for this dot
+        return failure();
+      }
+      // if k % 4 != 0: can not use DOT instruction, continue with FMA
+    }
+
+    // check that dot is not legalized already
+    if (isLegalFMAForm(dotOp)) {
+      return failure();
+    }
+
+    // Legalize dot for FMA case
+
+    // find common type, larger or equal of all operand types
+    SmallVector<Type> opElTy{aElTy, bElTy, cElTy, dElTy};
+    unsigned maxBitsize = 8;
+    for (auto elTy : opElTy)
+      maxBitsize = std::max(maxBitsize, elTy.getIntOrFloatBitWidth());
+    assert(maxBitsize <= 32);
+    Type commonTy =
+        maxBitsize <= 16 ? rewriter.getF16Type() : rewriter.getF32Type();
+
+    // check that type is compatible with all operands
+    // fallback to fp32 if not
+    if (commonTy.isF16()) {
+      for (auto elTy : opElTy) {
+        if (elTy.isInteger() && elTy.getIntOrFloatBitWidth() > 8) {
+          commonTy = rewriter.getF32Type();
+          break;
+        }
+        if (elTy.isBF16()) {
+          commonTy = rewriter.getF32Type();
+          break;
+        }
+      }
+    }
+
+    auto newA = castToElTy(rewriter, a, commonTy);
+    auto newB = castToElTy(rewriter, b, commonTy);
+    auto newC = castToElTy(rewriter, c, commonTy);
+
+    auto newDot = rewriter.create<DotOp>(dotOp.getLoc(), newC.getType(), newA,
+                                         newB, newC, dotOp.getInputPrecision(),
+                                         dotOp.getMaxNumImpreciseAcc());
+    auto newD = castToElTy(rewriter, newDot.getResult(), dElTy);
+    d.replaceAllUsesWith(newD);
+    dotOp.erase();
```

Comment:
Use `rewriter.replaceOp` instead of manually erase. Manually erase can cause the rewrite framework missing tracking and then weird issues.